### PR TITLE
Implemented Alter Table OPEN/CLOSE functionality

### PR DESCRIPTION
**Pull request by Autophagy**
`Schema.java` is being refactored in another PR, which I will update here when it is merged.

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 1]
```diff
@@ -107,7 +108,64 @@ public AlterTableOperation(ClusterService clusterService,
...
+                    }
+                }
+            }
+        } else {
+            if (analysis.openTable()) {
+                results.add(openTable(table.ident().indexName()));
+            } else {
+                results.add(closeTable(table.ident().indexName()));
+            }
+        }
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: this whole block can be simplified a lot:
```
String[] concreteIndices;
Optional<PartitionName> partitionName = analysis.partitionName();
if (partitionName.isPresent()) {
    concreteIndices = new String[]{partitionName.get().indexName()};
} else {
    concreteIndices = analysis.table().concreteIndices();
}
if (analysis.openTable()) {
.....
```

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 2]
```diff
@@ -107,7 +108,64 @@ public AlterTableOperation(ClusterService clusterService,
...
+
+        final CompletableFuture<Long> result = new CompletableFuture<>();
+        applyMultiFutureCallback(result, results);
+        return result;
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: this all above is not needed at all, you'll only have 1 future to return. so directly return the future of close/openTable()

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 3]
```diff
@@ -211,14 +269,17 @@ public void fail(@Nonnull Throwable t) {
         // merge settings
         Settings.Builder settingsBuilder = Settings.builder();
         settingsBuilder.put(indexTemplateMetaData.settings());
-        settingsBuilder.put(newSettings);
+        if (newSettings != null) {
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: how is this related to this PR?

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 4]
```diff
@@ -92,6 +92,11 @@ public DocTableInfo getDroppableTable(TableIdent tableIdent) {
             throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
                 "%s is an alias and hence not dropable.", tableInfo.ident()));
         }
+        if (docTableInfo.closed()) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "The table %s is closed. Write, Drop or Alter operations (other than OPEN) are not supported",
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: All operation beside ALTER .... OPEN are not supported

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 5]
```diff
@@ -106,6 +111,25 @@ public DocTableInfo getWritableTable(TableIdent tableIdent) {
             throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
                 "%s is an alias. Write, Drop or Alter operations are not supported", tableInfo.ident()));
         }
+        if (docTableInfo.closed()) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "The table %s is closed. Write, Drop or Alter operations (other than OPEN) are not supported",
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: same here

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 6]
```diff
@@ -106,6 +111,25 @@ public DocTableInfo getWritableTable(TableIdent tableIdent) {
...
+    public DocTableInfo getOpenableOrClosableTable(TableIdent tableIdent) {
+        TableInfo tableInfo = getTableInfo(tableIdent);
+        if (!(tableInfo instanceof DocTableInfo)) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "The table %s is read-only. Write, Drop or Alter operations are not supported", tableInfo.ident()));
+        }
+        DocTableInfo docTableInfo = (DocTableInfo) tableInfo;
+        if (docTableInfo.isAlias() && !docTableInfo.isPartitioned()) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "%s is an alias. Write, Drop or Alter operations are not supported", tableInfo.ident()));
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: ... Only READ operations are supported.

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 7]
```diff
@@ -106,6 +111,25 @@ public DocTableInfo getWritableTable(TableIdent tableIdent) {
...
+                tableInfo.ident()));
+        }
+        return docTableInfo;
+    }
+
+    public DocTableInfo getOpenableOrClosableTable(TableIdent tableIdent) {
+        TableInfo tableInfo = getTableInfo(tableIdent);
+        if (!(tableInfo instanceof DocTableInfo)) {
+            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                "The table %s is read-only. Write, Drop or Alter operations are not supported", tableInfo.ident()));
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: ... Only READ operations are supported.

File: sql/src/test/java/io/crate/integrationtests/OpenCloseTableIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+        IndexMetaData idx = client().admin().cluster().prepareState().execute().actionGet().getState().metaData().indices().get(indices[0]);
+
+        assertEquals(1, response.rowCount());
+        assertEquals(true, response.rows()[0][0]);
+        assertEquals(IndexMetaData.State.CLOSE, idx.getState());
+    }
+
+    @Test
+    public void testReopenTable() throws Exception {
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/OpenCloseTableIntegrationTest.java]: imho only this test is required as it does the same like the tests above

File: sql/src/test/java/io/crate/integrationtests/OpenCloseTableIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        execute("create table t (i int)");
+        ensureYellow();
+        String[] indices = client().admin().indices().getIndex(new GetIndexRequest()).actionGet().getIndices();
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/OpenCloseTableIntegrationTest.java]: don't think we need that, it is fixed to `doc.t`

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java (hunk 3 above)
Comment (Autophagy) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: It's not, I did't see that I could pass Settings.Empty into the function, my mistake.

File: sql/src/main/java/io/crate/metadata/Schemas.java (hunk 4 above)
Comment (Autophagy) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: These get addressed in the other PR, so I'll just remove the check from here, it is not relevant (doc table info has no closed)

File: sql/src/main/java/io/crate/metadata/Schemas.java (hunk 5 above)
Comment (Autophagy) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: as above

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 10]
```diff
@@ -107,7 +108,58 @@ public AlterTableOperation(ClusterService clusterService,
...
+
+        String[] concreteIndices;
+        Optional<PartitionName> partitionName = analysis.partitionName();
+        if (partitionName.isPresent()) {
+            concreteIndices = new String[]{partitionName.get().asIndexName()};
+        } else {
+            concreteIndices = analysis.table().concreteIndices();
+            if(table.isPartitioned()) {
+                HashMap<String, Object> metaMap = new HashMap<>();
+                metaMap.put("_meta", new HashMap(){{put("closed", true);}});
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: don't use double curly braces

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 11]
```diff
@@ -107,7 +108,58 @@ public AlterTableOperation(ClusterService clusterService,
...
+                if (analysis.openTable()) {
+                    //Remove the mapping from the template.
+                    results.add(updateTemplate(new HashMap<>(), metaMap, Settings.EMPTY, table.ident()));
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: use Collections.emptyMap()

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 12]
```diff
@@ -208,6 +267,10 @@ public void fail(@Nonnull Throwable t) {
         // merge mappings
         Map<String, Object> mapping = mergeTemplateMapping(indexTemplateMetaData, newMappings);
 
+        // remove mappings
+
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: remove empty line

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 13]
```diff
@@ -198,6 +250,13 @@ public void fail(@Nonnull Throwable t) {
     private CompletableFuture<Long> updateTemplate(Map<String, Object> newMappings,
                                                    Settings newSettings,
                                                    TableIdent tableIdent) {
+        return updateTemplate(newMappings, new HashMap<>(), newSettings, tableIdent);
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: use Collections.emptyMap()

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 14]
```diff
@@ -107,7 +108,58 @@ public AlterTableOperation(ClusterService clusterService,
...
+        List<CompletableFuture<Long>> results = new ArrayList<>(2);
+        DocTableInfo table = analysis.table();
+
+        String[] concreteIndices;
+        Optional<PartitionName> partitionName = analysis.partitionName();
+        if (partitionName.isPresent()) {
+            concreteIndices = new String[]{partitionName.get().asIndexName()};
+        } else {
+            concreteIndices = analysis.table().concreteIndices();
+            if(table.isPartitioned()) {
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: missing space

File: sql/src/main/java/io/crate/analyze/AlterTableOpenCloseAnalyzedStatement.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+    public AlterTableOpenCloseAnalyzedStatement(DocTableInfo tableInfo,
+                                                @Nullable PartitionName partitionName,
+                                                boolean openTable) {
+        this.tableInfo = tableInfo;
+        this.partitionName = partitionName;
+        this.openTable = openTable;
+    }
+
+    public DocTableInfo table() {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/AlterTableOpenCloseAnalyzedStatement.java]: pls rename to tableInfo

File: sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java [hunk 16]
```diff
@@ -107,7 +108,58 @@ public AlterTableOperation(ClusterService clusterService,
         return result;
     }
 
-    private class ResultSetReceiver implements ResultReceiver {
+    public CompletableFuture<Long> executeAlterTableOpenClose(final AlterTableOpenCloseAnalyzedStatement analysis) {
+        List<CompletableFuture<Long>> results = new ArrayList<>(2);
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: first check should be to check if any operation is needed at all (close closed table, open opened table). if no operation is needed, we can return an immediate future

Comment (Autophagy) [inline sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java]: In the case of an open partitioned table with a closed partition, how would we handle this? Since the table is already "open", yet opening it would open all closed partitions underneath it.

**Comment by Autophagy** [comment]
Ive refactored all these tests/valid operation checks in the other PR, so i'll reset my changes here.
